lib/net/ssh/transport/algorithms.rb

```diff
@@ -213,22 +213,32 @@ def prepare_preferred_algorithms!
         options[:compression] = %w(zlib@openssh.com zlib) if options[:compression] == true
 
         ALGORITHMS.each do |algorithm, list|
-          algorithms[algorithm] = list.dup
-
           # apply the preferred algorithm order, if any
           if options[algorithm]
-            algorithms[algorithm] = Array(options[algorithm]).compact.uniq
-            unsupported = []
-            algorithms[algorithm].select! do |name|
-              supported = ALGORITHMS[algorithm].include?(name)
-              unsupported << name unless supported
-              supported
-            end
-            lwarn { "unsupported #{algorithm} algorithm: `#{unsupported}'" } unless unsupported.empty?
-
-            if options[:append_all_supported_algorithms]
-              list.each { |name| algorithms[algorithm] << name unless algorithms[algorithm].include?(name) }
+            optionals = Array(options[algorithm]).compact.uniq
+
+            if optionals.first.start_with?('+')
```

Comment:
we should be still filtering the unsupported algs. We don't want to advertise something we don't support.